lang/java/compiler/src/main/java/org/apache/avro/compiler/specific/SpecificCompiler.java

```diff
@@ -660,13 +661,24 @@ public String javaUnbox(Schema schema) {
 
   public boolean hasLogicalTypeField(Schema schema) {
     for (Schema.Field field : schema.getFields()) {
-      if (field.schema().getLogicalType() != null) {
+      if (field.schema().getLogicalType() != null || anyUnionMemberHasLogicalType(field.schema())) {
         return true;
       }
     }
     return false;
   }
 
+  private boolean anyUnionMemberHasLogicalType(Schema schema) {
```

Comment:
I think this makes sense in the context of Avro 1.8 and the way it currently views unions/collections, but only because we are only ever using null and some other type and never a mixed type collection. Therefore I think this change would more or less constitute a true FTS extension here in that it is hard to predict how this issue will really be resolved for real in avro and therefore we may end up creating scenarios that are not going to be forwards compatible with whatever approach they take. Basically, they are likely going to have to resolve to the idea that a collection may not have a single type and then they have to make some sort of determination on what type to treat the value as for any given context. There are some discussions around this in 1.9 but I don't think they have come to any conclusions on it yet, but maybe they have.

So, I think for a change like this we need to add some inline documentation here around the nature and meaning of this change and what use case it is supporting, etc so we can be sure that when we upgrade to 1.9 we understand why this is here and how to deal with it. Really I think we need a test to be added in AVRO to validate the behavior we expect so that way it would be easier to validate correct behavior in the inevitable upgrade or when pulling in fixes from 1.9 into 1.8.

All that said, I wonder if this isn't something that should be fixed in the getLogicalType() method? It seems like the union schema should be able to answer this question and not the parent schema having to realize that the field schema is a union and then treating it differently?

Reply:
So this is the issue that was really puzzling me on Friday, and I now wish I had looked at the original thread where an Avro contributor pointed the flaw out before I had to find and fix this on my own: https://github.com/apache/avro/pull/118#issuecomment-322429757

The only place `hasLogicalTypeField` is called is in the Java compiler, where it determines whether or not to include the collection of logical type conversions in the generated Java class

Reply:
Basically the case this solves (and the case our current tests in avro-utils overlooked, need to add this) is when a Schema has only union logical types and no non-union logical types.

This schema will work fine without the change:
```json
{
         "namespace": "com.fanthreesixty.avro.testhelper.generated",
         "type": "record",
         "name": "SimpleUnionRecord",
         "fields": [
             {
                 "name": "fixed_union_field",
                 "type": ["null", "com.fanthreesixty.avro.testhelper.generated.Currency_BillionMax"],
                 "default": null
             },{
               "name": "fixed_field",
               "type": "com.fanthreesixty.avro.testhelper.generated.Currency_BillionMax"
             }
         ]
  }```

This schema requires the change as it has only a union logical type and the original method didn't cover that case:
```json
{
         "namespace": "com.fanthreesixty.avro.testhelper.generated",
         "type": "record",
         "name": "SimpleUnionRecord",
         "fields": [
             {
                 "name": "fixed_union_field",
                 "type": ["null", "com.fanthreesixty.avro.testhelper.generated.Currency_BillionMax"],
                 "default": null
             }
         ]
  }```

Reply:
That said, I'll add a comment and test case to cover this